src/token/cose/sign/mod.rs

```diff
@@ -0,0 +1,295 @@
+        &mut self,
+        alg: Algorithm,
+        key: &CoseEc2Key<'_, Self::Error>,
+        signature: &[u8],
+        target: &[u8],
+    ) -> Result<(), CoseCipherError<Self::Error>>;
+}
+
+fn is_valid_ecdsa_key<'a, BE: Display>(
+    algorithm: &Algorithm,
+    parsed_key: CoseParsedKey<'a, BE>,
+    key_should_be_private: bool,
+) -> Result<CoseEc2Key<'a, BE>, CoseCipherError<BE>> {
+    // Checks according to RFC 9053, Section 2.1 or RFC 8812, Section 3.2.
+
+    // Key type must be EC2
+    let ec2_key = if let CoseParsedKey::Ec2(ec2_key) = parsed_key {
+        ec2_key
+    } else {
+        return Err(CoseCipherError::KeyTypeAlgorithmMismatch(
+            parsed_key.as_ref().kty.clone(),
+            algorithm.clone(),
+        ));
+    };
+
+    // If algorithm in key is set, it must match our algorithm
+    if let Some(alg) = &ec2_key.as_ref().alg {
+        if alg != algorithm {
+            return Err(CoseCipherError::KeyAlgorithmMismatch(
+                alg.clone(),
+                algorithm.clone(),
+            ));
+        }
+    }
+
+    // Key must contain private key information to perform signature.
+    if key_should_be_private && ec2_key.d.is_none() {
+        return Err(CoseCipherError::MissingKeyParam(KeyParam::Ec2(
+            Ec2KeyParameter::D,
+        )));
+    } else if ec2_key.x.is_none() || ec2_key.y.is_none() {
+        return Err(CoseCipherError::MissingKeyParam(KeyParam::Ec2(
+            Ec2KeyParameter::X,
+        )));
+    }
+
+    Ok(ec2_key)
+}
+
+fn try_sign<'a, B: CoseSignCipher, CKP: CoseKeyProvider<'a>>(
+    backend: &mut B,
+    key_provider: &mut CKP,
+    protected: Option<&Header>,
+    unprotected: Option<&Header>,
+    tosign: &[u8],
+) -> Result<Vec<u8>, CoseCipherError<B::Error>> {
+    let parsed_key = determine_key_candidates(
+        key_provider,
+        protected,
+        unprotected,
+        &KeyOperation::Assigned(iana::KeyOperation::Sign),
+        false,
+    )?
+    .into_iter()
+    .next()
+    .ok_or(CoseCipherError::NoKeyFound)?;
+    let algorithm = determine_algorithm(&parsed_key, protected, unprotected)?;
+
+    let sign_fn = match algorithm {
+        Algorithm::Assigned(
+            iana::Algorithm::ES256
+            | iana::Algorithm::ES384
+            | iana::Algorithm::ES512
+            | iana::Algorithm::ES256K,
+        ) => {
+            // Check if this is a valid ECDSA key.
+            let ec2_key = is_valid_ecdsa_key::<B::Error>(&algorithm, parsed_key, true)?;
+
+            // Perform signing operation using backend.
+            move |tosign| return backend.sign_ecdsa(algorithm, &ec2_key, tosign)
+        }
+        v @ (Algorithm::Assigned(_)) => {
+            return Err(CoseCipherError::UnsupportedAlgorithm(v.clone()))
+        }
+        // TODO make this extensible? I'm unsure whether it would be worth the effort, considering
+        //      that using your own (or another non-recommended) algorithm is not a good idea anyways.
```

Comment:
Alright, in that case i might also make some more larger improvements throughout the code.
I am pretty sure some methods could then take `coset::iana::Algorithm` instead of `coset::Algorithm`, which might save some matching clauses.

Reply:
Yes, that sounds like a good idea.